### ISSUE TITLE
Allow libbpf-tools to run on old kernels

### ISSUE THREAD

**Issue by chenhengqi** [iovisor/bcc#4231]
We have a bunch of libbpf tools require features like `fentry` / `tp_btf` which only available on kernel v5.5+.
Running these tools on old kernels result in EINVAL. Instead of error out, we can fallback to kprobe.
Open this ticket to keep track of this process:

* [ ] biolatency
* [ ] biosnoop
* [ ] biostacks
* [ ] bitesize
* [ ] cachestat
* [ ] cpudist
* [ ] cpufreq
* [ ] drsnoop
* [x] hardirqs
* [ ] mdflush
* [ ] numamove
* [ ] offcputime
* [ ] readahead
* [x] softirqs

If you are interested, please feel free to pick it up.

**Comment by chenhengqi** [comment]
Using `raw_tp` is enough, this will make the tools running on older LTS kernels (like v4.19/v5.4).

**Comment by woodpenker** [comment]
Do we consider providing a cmd flag such as `-btf` to let user provide their BTF file for the kernel, so that we can use tools even on 4.x kernels?

**Comment by chenhengqi** [comment]
> Do we consider providing a cmd flag such as `-btf` to let user provide their BTF file for the kernel, so that we can use tools even on 4.x kernels?

We have such support at build time. See https://github.com/iovisor/bcc/pull/3889 .

**Comment by woodpenker** [comment]
> > Do we consider providing a cmd flag such as `-btf` to let user provide their BTF file for the kernel, so that we can use tools even on 4.x kernels?
> 
> We have such support at build time. See #3889 .

It seems to rely on the BTF hub and won't support self-built kernels or some old kernels which can't be found in the  hub, eg:  `Ubuntu 4.18.0-22`. Since `LIBBPF_OPTS` is supported, how about letting the user make a choice?
example:
```c
if (env.btf_path) {
  LIBBPF_OPTS(bpf_object_open_opts, bpf_opts, .btf_custom_path = env.btf_path);
  skel = xxx_bpf__open_opts(&bpf_opts);
} else {
  skel = xxx_bpf__open();
}
```

**Comment by chenhengqi** [comment]
Instead of touching every tool, I'd suggest providing the BTF path through environment variable and teach ensure_core_btf() to recognize it.

**Comment by woodpenker** [comment]
> Instead of touching every tool, I'd suggest providing the BTF path through environment variable and teach ensure_core_btf() to recognize it.

Thanks for explaining, I may try that. But I realize another problem with global values, only legacy codes are supported on old kernels like [this](https://github.com/libbpf/libbpf-bootstrap/blob/a7c0f7e4a286c6c8ba39e8fc736609412fa2de43/examples/c/minimal_legacy.bpf.c#L14). To support them, maybe we need some mechanisms to auto-generate legacy codes instead of global values?

**Comment by woodpenker** [comment]
BTW, some tools such as [softirqs](https://github.com/iovisor/bcc/blob/7c02284faa/libbpf-tools/softirqs.c) seems not using `ensure_core_btf()`, is it intended?

**Comment by chenhengqi** [comment]
Yes, `tp_btf` relies on BPF trampoline which support `/sys/kernel/btf/vmlinux` only.

**Comment by woodpenker** [comment]
> Yes, `tp_btf` relies on BPF trampoline which support `/sys/kernel/btf/vmlinux` only.

Thanks, I want to try to teach some tools listed above to fall back to raw_tp or Kprobe, maybe start from `offcputime` which I am familiar with. Another question, it seems kernel support auto falls back to tp now after this [patch](https://lore.kernel.org/bpf/20220421033945.3602803-3-andrii@kernel.org/), am I right?

**Comment by chenhengqi** [comment]
> Another question, it seems kernel support auto falls back to tp now after this [patch](https://lore.kernel.org/bpf/20220421033945.3602803-3-andrii@kernel.org/), am I right?

No.

**Comment by woodpenker** [comment]
Since `cpufreq` need Memory-mapping BPF maps support starting from v5.7([b97e12e594eb](https://github.com/torvalds/linux/commit/b97e12e594eb3d0fba35c99a83fef56ff692048b) & [333291ce5055](https://github.com/torvalds/linux/commit/333291ce5055f2039afc907badaf5b66bc1adfdc)), I think there is no need to change it.
